subscription-engine/kafka-ws/index.js

```diff
@@ -1,60 +1,93 @@
+
+const websocketTarball = require('./app/WebsocketTarball');
 
-Logger.setVerbose(logConfig.log_verbose);
+const terminus = require('./app/Terminus');
 
 const logger = new Logger();
 
+let server = null;
+
+if (serverCfg.tls) {
+  logger.info('Initializing the HTTP server (Using TLS Protocol)...');
+  const options = {
+    cert: fs.readFileSync(serverCfg.tls_cert_file),
+    key: fs.readFileSync(serverCfg.tls_key_file),
+    ca: [fs.readFileSync(serverCfg.tls_ca_file)],
+    rejectUnauthorized: true,
+    requestCert: true,
+  };
+  server = https.createServer(options, application.expressApp);
+} else {
+  logger.info('Initializing the HTTP server...');
+  server = http.createServer(application.expressApp);
+}
+
+/* Configures the application's HTTP and WS routes */
+application.configure(server);
+
+server.listen(serverCfg.port, serverCfg.host, async () => {
+  logger.info('HTTP server is ready to accept connections!');
+  logger.info(server.address());
+
+  // Initializes the sticky tarball
+  await websocketTarball.init();
+});
+
+/* adds health checks and graceful shutdown to the application */
+terminus.setup(server);
+
 const unhandledRejections = new Map();
+
 // the unhandledRejections Map will grow and shrink over time,
 // reflecting rejections that start unhandled and then become handled.
 process.on('unhandledRejection', (reason, promise) => {
   // The 'unhandledRejection' event is emitted whenever a Promise is rejected and
   // no error handler is attached to the promise within a turn of the event loop.
   logger.error(`Unhandled Rejection at: ${reason.stack || reason}.`);
+
   unhandledRejections.set(promise, reason);
+
   logger.debug(`unhandledRejection: List of Unhandled Rejection size ${unhandledRejections.size}`);
 });
+
 process.on('rejectionHandled', (promise) => {
   // The 'rejectionHandled' event is emitted whenever a Promise has
   // been rejected and an error handler was attached to it
   // later than one turn of the Node.js event loop.
   logger.debug('rejectionHandled: A event');
+
   unhandledRejections.delete(promise);
+
   logger.debug(`rejectionHandled: List of Unhandled Rejection size ${unhandledRejections.size}`);
 });
 
 process.on('uncaughtException', async (ex) => {
   // The 'uncaughtException' event is emitted when an uncaught JavaScript
   // exception bubbles all the way back to the event loop.
   logger.error(`uncaughtException: Unhandled Exception at: ${ex.stack || ex}. Bailing out!!`);
-  // TODO: stop server (connection redis, kafka consumer, etc.)
-  process.exit(1);
-});
 
-// init server
-try {
-  const server = new Server();
-  (async () => {
-    await server.init();
-  })().catch((error) => {
-    logger.error(`Caught an error when trying init Server ${error}`);
-    throw error;
-  });
-} catch (error) {
-  logger.error(`Caught a final error: ${error.stack || error}`);
   // TODO: stop server (connection redis, kafka consumer, etc.)
   process.exit(1);
```

Comment:
If I choose the solution `process.exitCode = 1;` it seems to me that i would have to throw another `uncaughtException` to try to terminate the process, but we are inside a handler to catch it (`process.on('uncaughtException')`)... can you see the infinite loop? (Just kidding).

I honestly didn't understand the reason for catching an `uncaughtException` and immediately force the process to end. The logger probably won't have time to write something to a file (if that's the idea).

I prefer to let the application crash and let the docker responsible for up a new container:
(https://shapeshed.com/uncaught-exceptions-in-node/)

it depends on me, I remove everything below the line 61, but do not know why this code was written, much less what would be the impact of this removal.

Reply:
As the docs says:
```
Rather than calling process.exit() directly, the code should set the process.exitCode and allow the process to exit naturally by avoiding scheduling any additional work for the event loop
```

I think this would not throw some exception or anything like that. But since we don't know how this really works, I think it's fair to keep the `process.exit(1)`.

Reply:
OK, I think I misunderstood how the exitCode works.